tests/unittests/test_utils.py

```diff
@@ -0,0 +1,269 @@
+            },
+            id="noargs_raise_valueerror_with_no_msg",
+        ),
+        pytest.param(
+            {
+                "init_exceptions_dict": {
+                    KeyError: [],
+                },
+                "runtime_exception": ValueError(),
+            },
+            {
+                "raises": ValueError,
+            },
+            id="init_keyerror_raise_valueerror_with_no_msg",
+        ),
+        pytest.param(
+            {
+                "init_exceptions_dict": {
+                    ValueError: ["test"],
+                },
+                "runtime_exception": ValueError("test"),
+            },
+            {
+                "exception_log_regex": "^.*\nLast exception: ValueError: test$",
+            },
+            id="init_valueerror_with_msg_raise_valueerror_with_allowed_msg",
+        ),
+        pytest.param(
+            {
+                "init_exceptions_dict": {
+                    ValueError: ["allowed exception text"],
+                },
+                "runtime_exception": ValueError("test"),
+            },
+            {
+                "raises": ValueError,
+            },
+            id="init_valueerror_with_msg_raise_valueerror_with_invalid_msg",
+        ),
+        pytest.param(
+            {
+                "init_exceptions_dict": {
+                    KeyError: ["allowed exception text"],
+                    IndexError: ["allowed exception text"],
+                    ValueError: ["allowed exception text"],
+                },
+                "runtime_exception": IndexError("my allowed exception text"),
+            },
+            {
+                "exception_log_regex": "^.*\nLast exception: IndexError: my allowed exception text$",
+            },
+            id="init_multi_exceptions_raise_allowed_with_allowed_msg",
+        ),
+        pytest.param(
+            {
+                "init_exceptions_dict": {
+                    KeyError: ["allowed exception text"],
+                    IndexError: ["allowed exception text"],
+                    ValueError: ["allowed exception text"],
+                },
+                "runtime_exception": IndexError("test"),
+            },
+            {
+                "raises": IndexError,
+            },
+            id="init_multi_exceptions_raise_allowed_with_invalid_msg",
+        ),
+    ],
+)
+def test_timeout_sampler_raises(test_params, expected):
+    def _raise_exception(runtime_exception):
+        if runtime_exception:
+            raise runtime_exception
+
+    def _run_timeout_sampler():
+        for _ in TimeoutSampler(
+            wait_timeout=1,
+            sleep=1,
+            func=_raise_exception,
+            exceptions_dict=test_params.get("init_exceptions_dict"),
+            print_log=False,
+            runtime_exception=test_params.get("runtime_exception"),
+        ):
+            continue
+
+    if expected.get("raises"):
```

Comment:
I would split it into 2 tests.
One that handles `raises` and one for `TimeoutExpiredError` exception check.